BJS2-40653 Стажировка 2/5 (Version 1.0)

Darkselect opened the pull request:
1. Add method with validator
2. Update method with handler

Darkselect committed:
BJS2-40653 Стажировка 2/5 (Version 1.0)

1. Created add method with validator
2. Created update method with handler

drewbal commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,42 @@
+public class InternshipController {
+    private final InternshipService internshipService;
+
+    @PostMapping("/create")
```
В Rest в названии пути принято использовать существительные. Также в твоем случае можно не указывать продолжение пути, т.е. Post запрос у тебя один, и вполне логично, что он выполняется в корневом пути

Darkselect replied:
Исправил

drewbal commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,42 @@
+        return internshipService.create(internShipCreatedDto);
+    }
+
+    @PutMapping("/update")
```
Аналогично предыдущему запросу. Можно не указывать и просто оставить @PutMapping.

Darkselect replied:
Исправил

drewbal commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,42 @@
+    private final InternshipService internshipService;
+
+    @PostMapping("/create")
+    public InternShipCreatedDto create(InternShipCreatedDto internShipCreatedDto) {
```
Если ты используешь аннотации для путей запросов, то тебе желательно поставить аннотации, чтобы входящий JSON конвертировался в dto, а также валидировался. То есть будет вот так:
`public InternShipCreatedDto create(@Vaild @RequestBody InternShipCreatedDto internShipCreatedDto)`

Эти вещи по идее будут поясняться в открывшихся лекциях, так что можно посмотреть их, или пока не использовать аннотации в этой задаче (хотя уверен, что ты посмотришь :) )

Darkselect replied:
Добавил

drewbal commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,42 @@
+
+    @PutMapping("/update")
+    public InternShipUpdatedDto update(InternShipUpdatedDto InternShipUpdatedDto) {
+        return internshipService.updatedDto(InternShipUpdatedDto);
```
аналогично

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipCreatedDto.java:
```diff
@@ -0,0 +1,44 @@
+public class InternShipCreatedDto {
+    private Long projectId;
+
+    @NotNull(message = "Mentor is can not be null")
```
Вот как раз для создания новой записи тебе нужен id с null, т.к. иначе БД не присвоит ему новый номер. И вообще в дто для создания новой записи можно не просить id, он тебе не нужен

Darkselect replied:
Убрал Null c ментора

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipCreatedDto.java:
```diff
@@ -0,0 +1,44 @@
+    private TeamMember mentorId;
+
+    @NotNull(message = "Start date can not be null")
+    @Builder.Default()
```
а что делает эта аннотация?

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipCreatedDto.java:
```diff
@@ -0,0 +1,44 @@
+    @Builder.Default()
+    private LocalDateTime startDate = LocalDateTime.now();
+
+    private LocalDateTime endDate = LocalDateTime.now();
```
дата окончания может быть null? Насколько я понимаю по разнице между датами проверку надо делать

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipCreatedDto.java:
```diff
@@ -0,0 +1,44 @@
+    private String description;
+
+    @NotBlank(message = "Internship name can not be blank")
+    @NotNull(message = "Internship name can not be null")
```
аннотация @NotNull уже исключает null, так что вторую можно дропнуть

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipCreatedDto.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Positive(message = "Internship duration can not be negative")
+    @NotNull(message = "Internship duration can not be null")
+    private Long createdBy;
```
название переменной лучше сделать createdById, чтобы было понятнее, а то не всегда виден тип данных и не понятно, что там. Вдруг имя?

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipUpdatedDto.java:
```diff
@@ -0,0 +1,47 @@
+@Data
+@Builder
+public class InternShipUpdatedDto {
+    @PositiveOrZero
```
эта аннотация допускает null и 0. Тебе нужно обязательно быть уверенным, что id существующей записи @NotNull, а также @Positive, т.к. не может быть равен 0

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipUpdatedDto.java:
```diff
@@ -0,0 +1,47 @@
+    @PositiveOrZero
+    private Long id;
+
+    private Long projectId;
```
наверное тоже @NotNull + @Positive? Стажировка же обязательно имеет реальный проект?

drewbal commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternShipUpdatedDto.java:
```diff
@@ -0,0 +1,47 @@
+    private InternshipStatus status;
+
+    @NotNull(message = "Internship owner cannot be null")
+    private Long createdBy;
```
createdById + @Positive

drewbal commented on src/main/java/faang/school/projectservice/handler/InternshipCompletionHandler.java:
```diff
@@ -0,0 +1,51 @@
+public class InternshipCompletionHandler {
+    private final Random random = new Random();
+
+    public void handleInternsCompletion(Internship internship) {
```
а не надо проверить, что internship не null и не порушит все?

RuslanZinovyev replied:
Дониер, мы закрываем комментарии в 2ух случаях:
1) Замечение другого разработчика адресовано
2) Оставлен понятный комментарий почему оно было проигнорировано

Пример:
В данном случае null сюда не прилетит, если стажировки не будут найдены в репозитории то вылетит EntityNotFound exception, задолго до того, как этот метод будет исполнен.

drewbal commented on src/main/java/faang/school/projectservice/handler/InternshipCompletionHandler.java:
```diff
@@ -0,0 +1,51 @@
+        }
+    }
+
+    public void internsToDismissal(List<@Positive Long> interns) {
```
а для чего эта аннотация @Positive в этом методе? Я думал она только для валидации на уровне контроллера нужна

Darkselect replied:
Чтобы не было отрицательных id

RuslanZinovyev replied:
Ты уже сделал эту валидацию на уровне DTO, тут она не требуется.

drewbal commented on src/main/java/faang/school/projectservice/handler/InternshipCompletionHandler.java:
```diff
@@ -0,0 +1,51 @@
+
+    public void internsToDismissal(List<@Positive Long> interns) {
+        if (interns != null && !interns.isEmpty()) {
+            interns.forEach(interns::remove);
```
а не проще очистить список через interns.clear()?

Darkselect replied:
мне кажется одно и тоже

drewbal replied:
Почти, но только .clear() удалит все за 1 действие, а forEach пойдет удалять каждый и сделает n действий

drewbal commented on src/main/java/faang/school/projectservice/handler/InternshipCompletionHandler.java:
```diff
@@ -0,0 +1,51 @@
+    private boolean hasCompletedAllTasks(TeamMember intern) {
+        return intern.getStages().stream()
+                .flatMap(stage -> stage.getTasks().stream())
+                .allMatch(task -> task.getStatus() == TaskStatus.DONE);
```
Этот метод просто красота!!

drewbal commented on src/main/java/faang/school/projectservice/handler/InternshipCompletionHandler.java:
```diff
@@ -0,0 +1,51 @@
+
+    private void assignNewRole(TeamMember intern) {
+        intern.getRoles().remove(TeamRole.INTERN);
+        intern.getRoles().add(TeamRole.values()[random.nextInt(TeamRole.values().length - 1)]);
```
за смекалку 5! Но ты уверен, что в реальных проектах должность стажеру назначается рандомом? :)))

Darkselect replied:
А они в условии не сказали как именно распределять роль после прохождения стажировки

drewbal replied:
А здравая логика что подсказывает? Думаешь будут пользоваться нашим приложением, если оно ведет себя неконтролируемо?

Тут вообще некорректно метод сделан, потому что надо передавать в аргументы интерна и его новую роль, а внутри метода убирать старую и присваивать новую. Или как ты думаешь?

drewbal commented on src/main/java/faang/school/projectservice/mapper/InternShipUpdateMapper.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.stream.Collectors;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface InternShipUpdateMapper {
```
я бы тебе предложил сделать единый маппер на все дто. Ведь по сути ты одни и те же поля маппишь на одну и ту же сущность. А в зависимости от набора полей в дто маппер будет все правильно делать. Нет поля - он и не маппит его просто

drewbal commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,42 @@
+
+    @PostMapping("/create")
+    public InternShipCreatedDto create(InternShipCreatedDto internShipCreatedDto) {
+        return internshipService.create(internShipCreatedDto);
```
у нас сущность называется Internship, поэтому лучше и созданные инструменты типа дто называть соответствующе, т.е. internshipCreatedDto

drewbal commented on src/main/java/faang/school/projectservice/validator/InternshipDurationValidator.java:
```diff
@@ -0,0 +1,24 @@
+@Slf4j
+@Component
+public class InternshipDurationValidator {
+    public void durationValidate(InternShipCreatedDto internShipCreatedDto) {
```
название метода начинается с глагола, т.е. будет vallidateDuration

drewbal commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,45 @@
+public class InternshipService {
+    private final ProjectService projectService;
+    private final InternshipRepository internshipRepository;
+    private final InternShipCreateMapper internShipCreateMapper;
```
в англ слово "стажировка" пишется как единое Internship, Ship не является отдельным словом. Лучше делать классы, переменные и методы, следуя этому паттерну

drewbal commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,45 @@
+    private final InternshipRepository internshipRepository;
+    private final InternShipCreateMapper internShipCreateMapper;
+    private final InternShipUpdateMapper internShipUpdateMapper;
+    private final InternshipDurationValidator internShipDurationValidator;
```
Хаха, а вот тут название класса у тебя корректно, а в переменной снова вылез Ship :))

drewbal commented on src/main/java/faang/school/projectservice/mapper/InternShipCreateMapper.java:
```diff
@@ -0,0 +1,39 @@
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface InternShipCreateMapper {
+    @Mapping(source = "projectId", target = "project")
+    @Mapping(source = "mentorId", target = "mentorId")
```
ты уверешь, что хочешь смаппить Long projectId на Project project? ты перетрешь сущность. В случае, если в дто примитив, а в сущности объект, то нужно игнорить это поле.

drewbal commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,45 @@
+    public InternShipCreatedDto create(InternShipCreatedDto internShipCreatedDto) {
+        internShipDurationValidator.durationValidate(internShipCreatedDto);
+        projectService.getProjectTeamMembersIds(internShipCreatedDto);
+        return internShipCreateMapper.toDto(internshipRepository.save(internShipCreateMapper.toEntity(internShipCreatedDto)));
```
Перед сохранением сущности было бы неплохо восстановить ее из дто, добавив те параметры, которые ты мог игнорить маппером

drewbal commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,45 @@
+
+        completionHandler.internsToDismissal(internShipUpdatedDto.getInternToDismissal());
+
+        if (internShipUpdatedDto.getStatus() == InternshipStatus.COMPLETED) {
```
а вот такую проверку лучше вынести в валидатор, чтобы сервисный метод занимался только операциями по вызову/сохранению данных

Darkselect replied:
Убрал

drewbal commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,45 @@
+            completionHandler.handleInternsCompletion(internship);
+        }
+
+        Internship savedInternship = internshipRepository.save(internShipUpdateMapper.toEntity(internShipUpdatedDto));
```
если ты просто сохранишь такую смапленную дто, то потрешь много инфы в БД. У тебя маппится Long на Project

Darkselect replied:
Исправил

drewbal commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,35 @@
+
+    public void getProjectTeamMembersIds(InternShipCreatedDto internShipCreatedDto) {
+        Long projectId = internShipCreatedDto.getProjectId();
+        Project projectById = projectRepository.getProjectById(projectId);
```
чуток странное название переменной. почему бы не просто project?

Darkselect replied:
Ага

RuslanZinovyev replied:
Переименуй в просто project

drewbal commented on src/main/java/faang/school/projectservice/validator/InternshipDurationValidator.java:
```diff
@@ -0,0 +1,24 @@
+        LocalDateTime startDate = internShipCreatedDto.getStartDate();
+        LocalDateTime endDate = internShipCreatedDto.getEndDate();
+
+        if (ChronoUnit.MONTHS.between(startDate, endDate) > 3 ) {
```
и вот поэтому в дто лучше сделать @NotNull для endDate

drewbal reviewed (commented):
Все норм, есть несколько моментов, на которые обратить внимание. А пара методов вообще красотища! :)

Только тесты не успеваю посмотреть

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 2.0)

1. Created add method with validator
2. Created update method with handler
3. Created filter method
4. Created get all internships method
5. Created find internship by id method

Without tests.

DeemaP commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,45 @@
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/internship")
```
Юра нам посоветовал использовать версионирование в пути контроллера)
@RequestMapping("/api/v1/intership")
И скинул эту статейку
https://habr.com/ru/companies/superjob/articles/577650/

RuslanZinovyev replied:
Дониер, если ты закрываешь комментарий, то ты должен либо адресовать его, либо оставить комментарий почему этого сделано не было. Мы сейчас начали использовать версионность в проекте и если тебе лень читать статью, либо смысл ее не понятен давай я попробую тезисно изложить преимущества такого подхода.

Обратная совместимость и гибкость, это озволяет в новой версии API добавлять новые функции не нарушая и не изменяя стабильные старые. Пользователи могут пользоваться стабильной версией если новая в активной разработке.

RuslanZinovyev replied:
Посмотри последние изменения Юры в его контроллерах, /api/v1 можно вынести в путь контекста конфигурации

DeemaP commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternshipUpdatedDto.java:
```diff
@@ -0,0 +1,46 @@
+    @Builder.Default
+    private LocalDateTime startDate = LocalDateTime.now();
+
+    @NotNull(message = "Start date can not be null")
```
В сообщении пишется о StartDate

RuslanZinovyev replied:
Комментарий выше проигнорирован

Darkselect replied:
Объяснил ситуацию. Ветка была поломана, сделал откат

DeemaP commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,45 @@
+package faang.school.projectservice.controller;
+
+import faang.school.projectservice.dto.client.internShip.InternshipCreatedDto;
```
переименуй пакет internship, чтоб без большой буквы)

RuslanZinovyev replied:
Мы не используем camelCase в названиях пакетов, пакет должен быть исключительно в низком регистре

DeemaP commented on src/main/java/faang/school/projectservice/validator/InternshipDurationValidator.java:
```diff
@@ -0,0 +1,24 @@
+        LocalDateTime startDate = internShipCreatedDto.getStartDate();
+        LocalDateTime endDate = internShipCreatedDto.getEndDate();
+
+        if (ChronoUnit.MONTHS.between(startDate, endDate) > 3 ) {
```
Эту тройку лучше вынести в константу, и заодно в логировании ее использовать

DeemaP commented on src/main/java/faang/school/projectservice/validator/InternshipDurationValidator.java:
```diff
@@ -0,0 +1,24 @@
+        if (ChronoUnit.MONTHS.between(startDate, endDate) > 3 ) {
+            throw new InternshipDurationException("Internship duration cannot exceed 3 months");
+        } else {
+            log.info("End of Internship");
```
Не совсем понял текст этого лога, здесь ты проверяешь, идет ли стажировка больше трех месяцев. Если больше - бросаешь исключение, если меньше - окончание стажировки.
То есть при создании стажировки у тебя в логах будет написано "Окончание стажировки". Наверное здесь лучше было бы писать что-то в формате "Стажировка идет меньше {} месяцев"

Но возможно я не корректно понял логику.

RuslanZinovyev replied:
Логика мне не понятна, ты можешь уйти в else если ты еще в процессе стажировки и почему при этом выходит сообщение в логе "End of Internship" не в полной мере ясно.

DeemaP commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternshipGetByIdDto.java:
```diff
@@ -0,0 +1,14 @@
+package faang.school.projectservice.dto.client.internShip;
```
Расскажи пожалуйста, а зачем так много разных dto для стажировок, интересен твой ход мыслей? 
Вот в этой, например, ты пользователю вернешь просто id. Тут даже имени нет и описания.
Почему не возвращать пользователю остальные данные?

Если пользователь запрашивает все стажировки, ты ему возвращаешь много данных, там имя, описание, дата начала, ментор)

Мне кажется ты мог все дто объединить в одну, только для обновления использовать отдельную дто)

апдейт:
для фильтрации тоже отдельная дто, я понял.

Darkselect replied:
Рассказал на созвоне

DeemaP commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternshipCreatedDto.java:
```diff
@@ -0,0 +1,44 @@
+    @Builder.Default()
+    private LocalDateTime startDate = LocalDateTime.now();
+
+    private LocalDateTime endDate = LocalDateTime.now();
```
Расскажи пожалуйста, очень интересно как это работает. 
Я о таком не знаю и использовал бы присвоение даты закрытия вручную в методе сервиса.
Может тут крутая фишка какая-то)

RuslanZinovyev replied:
Как будет работать твоя валидация если у тебя endDate создается тем же числом что и startDate? 

@NotNull(message = "End date cannot be null")
 private LocalDateTime endDate;

DeemaP commented on src/main/java/faang/school/projectservice/filter/InternShipStatusFilter.java:
```diff
@@ -0,0 +1,20 @@
+
+    @Override
+    public Stream<Internship> apply(Stream<Internship> dataStream, InternshipFilterDto filter) {
+        return dataStream.filter(internship -> internship.getStatus().equals(filter.getInternshipStatus()));
```
для сравнения отдельных элементов enum лучше использовать ==, это быстрее, чем вызывать метод contains.

DeemaP commented on src/main/java/faang/school/projectservice/mapper/InternshipCreateMapper.java:
```diff
@@ -0,0 +1,39 @@
+    @Mapping(source = "mentorId.id", target = "mentorId")
+    InternshipCreatedDto toDto(Internship entity);
+
+    default Project mapProject(Long projectId) {
```
Подскажи, мне еще не попадались просто маперы, где нужно было так делать.
Эти методы не нужно указывать внутри аннотации @Mapping? Маппер сам поймет что ему надо?

Darkselect replied:
Потому что может потребоваться преобразовать ID ментора в Объект TeamMember

DeemaP commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternshipFilterDto.java:
```diff
@@ -0,0 +1,26 @@
+@AllArgsConstructor
+@Builder
+public class InternshipFilterDto {
+    private Long id;
```
А зачем тебе в фильтре айди, если ты не фильтруешь по айди?

RuslanZinovyev replied:
@Darkselect Я не вижу что id был удален в итоге.

Darkselect replied:
у меня ветка была поломана, я сделал откат и заново все писал, и забыл сорри

DeemaP commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,109 @@
+        return internshipUpdateMapper.toDto(savedInternship);
+    }
+
+    public List<InternshipFilterDto> filterInternship(InternshipFilterDto filters) {
```
Для чего ты делал приватные методы? Влад говорил, что поход в базу - дело не дешевое, а тут ты дважды ходишь в базу, да еще и за списком всех элементов сразу. Наверное стоит как-то в один метод все сунуть, чтобы работать с одним списком и не ходить в базу дважды. 

Не могу пока быстро придумать как бы сразу вытаскивать статусы из TeamMember, но думаю, что это можно оптимизировать)

Darkselect replied:
Пока не нашел круче идеи

RuslanZinovyev replied:
Идея очень простая, сходить в базу один раз и уже работать с этим списком стажировок как тебе захочется.

Darkselect replied:
Реализовал

DeemaP reviewed (commented):
Привет! Прочитал код, оставил вопросы и комментарии)
Хорошая работа, но местами перемудрил, мне кажется)

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 3.0)

1. Created add method with validator
2. Created update method with handler
3. Created filter method
4. Created get all internships method
5. Created find internship by id method

With all tests.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+
+    public List<InternshipFilterDto> filterInternship(InternshipFilterDto internShipFilterDto) {
```
Название метода не верно, если контроллер должен получить все стажировки по фильтру то и название должно быть `getAllInternshipsByFilter()`

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternshipUpdatedDto.java:
```diff
@@ -0,0 +1,46 @@
+@AllArgsConstructor
+@Data
+@Builder
+public class InternshipUpdatedDto {
```
Зачем нам тут столько полей если ты их не используешь. Посмотри логику своего сервиса, все что тебе нужно Id, Status и список стажеров на исключение.

```
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class InternshipUpdatedDto {
    private Long id;

    @NotNull(message = "Internship status cannot be null")
    private InternshipStatus status;

    private List<TeamMember> internsToDismiss;
}
```

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/client/internShip/InternshipGetAllDto.java:
```diff
@@ -0,0 +1,29 @@
+@NoArgsConstructor
+@AllArgsConstructor
+public class InternshipGetAllDto {
+    private Long id;
```
Тут надо подумать, но возможно этой информации будет достаточно

```
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class InternshipGetAllDto {
    private Long id;
    private String name;
    private InternshipStatus status;
}
```

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,101 @@
+    private final ProjectService projectService;
+    private final InternshipRepository internshipRepository;
+    private final InternshipMapper internshipMapper;
+    private final InternshipDurationValidator internshipDurationValidator;
```
Не надо имя класса посвящать одной операции, назови просто` InternshipValidator` и там уже будет метод, который будет валидировать `duration`.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,35 @@
+
+        TeamMember mentorId = internShipCreatedDto.getMentorId();
+
+        if (!projectServiceValidator.isMentorPresent(teamMembersId, mentorId.getId())) {
```
сделай `projectServiceValidator.validateMentorPresence(teamMembersId, mentorId.getId());` и всю лонику валидации убери в валидатор, если у тебя уже есть валидатор, то нет смысла перегружать код сервиса ненужными условиями. Все проверки и исключения только в самом валидаторе.

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 4.0)

1. Created add method with validator
2. Created update method with handler
3. Created filter method
4. Created get all internships method
5. Created find internship by id method

With all tests.

Darkselect committed:
Merge branch 'Dragon-master-stream7' of https://github.com/CorporationX/project_service into Dragon-master-stream7-feature-BJS2-40653

# Conflicts:
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 6.0)
Fixed all

With all tests.

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 8.0)
Fixed all

With all tests.

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 7.0)
Fixed all

With all tests.

Darkselect committed:
Merge remote-tracking branch 'origin/Dragon-master-stream7-feature-BJS2-40653' into Dragon-master-stream7-feature-BJS2-40653

Darkselect committed:
Merge branch 'Dragon-master-stream7' of https://github.com/CorporationX/project_service into Dragon-master-stream7-feature-BJS2-40653

# Conflicts:
#	src/main/java/faang/school/projectservice/service/ProjectService.java

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 8.0)
Fixed all

With all tests.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/internship/InternshipCreatedDto.java:
```diff
@@ -0,0 +1,45 @@
+
+    @NotBlank(message = "Internship name can not be blank")
+    @NotNull(message = "Internship name can not be null")
+    private String name;
```
> NotBlank can be applied only to text values, and validates that the property isn’t null or whitespace.

Если ты повесил аннотацию NotBlank то аннотация NotNull в данном случае становится избыточной.

Darkselect replied:
Убрал

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/internship/InternshipFilterDto.java:
```diff
@@ -0,0 +1,24 @@
+@AllArgsConstructor
+@Builder
+public class InternshipFilterDto {
+    @Enumerated(EnumType.STRING)
```
Зачем здесь это аннотация? Ты планируешь DTO в базе сохранять? Ты понимаешь значение этой аннотации?

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/internship/InternshipFilterDto.java:
```diff
@@ -0,0 +1,24 @@
+    @NotNull(message = "Internship status can not be null")
+    private InternshipStatus internshipStatus;
+
+    @Enumerated(EnumType.STRING)
```
Зачем здесь это аннотация? Ты планируешь DTO в базе сохранять? Ты понимаешь значение этой аннотации?

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/internship/InternshipGetAllDto.java:
```diff
@@ -0,0 +1,22 @@
+    private Long id;
+
+    @NotNull
+    private String name;
```
NotBlank возможно более уместно тут

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/internship/InternshipUpdatedDto.java:
```diff
@@ -0,0 +1,28 @@
+@Builder
+public class InternshipUpdatedDto {
+    @PositiveOrZero
+    private Long id;
```
У тебя id может быть равен нулю?

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/internship/InternshipUpdatedDto.java:
```diff
@@ -0,0 +1,28 @@
+    @NotNull(message = "Internship status can not be null")
+    private InternshipStatus status;
+
+    private List<@Positive Long> interns;
```
Можно еще добавить NotEmpty

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,39 @@
+public class InternshipController {
+    private final InternshipService internshipService;
+
+    public InternshipCreatedDto createInternship(InternshipCreatedDto internShipCreatedDto) {
```
Если не добавить аннотацию Valid для всех принимающих DTO в качестве аргумента ендпоинтов, то валидационные правила что ты настроил на своих DTO не будут работать.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/handler/InternshipCompletionHandler.java:
```diff
@@ -0,0 +1,56 @@
+        }
+    }
+
+    public void handleInternsCompletion(Internship internship) {
```
Если этот метод используется как вспомогательный для работы `processInternshipCompletion` он должан быть приватным и спущен в самый низ класса.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/validator/InternshipValidator.java:
```diff
@@ -0,0 +1,26 @@
+        long internshipDuration = ChronoUnit.MONTHS.between(startDate, endDate);
+        if (internshipDuration > maxDurationMonths) {
+            throw new InternshipDurationException("Internship duration cannot exceed " + maxDurationMonths + " months");
+        } else {
```
Я думаю  тут else можно вообще убрать, информация о позитивном сценарии нам не интересна, только лог будет зря засоряться. Валидатор пишет что-то, если что-то пошло не так. Можно перед выбросом исключения добавить log.error("Сообщение о том что пошло не так")

Darkselect replied:
Оставил смайлик, надеюсь видно

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -121,5 +125,21 @@ private List<Project> getAllAccessibleProjects(Long currentUserId) {
                 .toList();
     }
+
+    public void getProjectTeamMembersIds(InternshipCreatedDto internShipCreatedDto) {
```
Очень странный метод, название говорит о том что мы должны что-то получить, но он ничего не возвращает и работает как валидационный. Исправь. Если это валидация то перенеси в посвященный класс валидатор, если он должен что-то вернуть согласено своему имени то пусть возвращает, иначе в сервисе ему делать нечего.

RuslanZinovyev replied:
Также никогда не добавляй публичные методы после приватных. Убедись что все твои классы сперва содержат публичные методы и только потом приватные.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,97 @@
+
+    public InternshipCreatedDto createInternship(InternshipCreatedDto internShipCreatedDto) {
+        internshipValidator.durationValidate(internShipCreatedDto);
+        projectService.getProjectTeamMembersIds(internShipCreatedDto);
```
Не уверен что конкретно делает данный вызов, если валидирует что-то то имя метода говорит об обратном. Это нужно исправить.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,97 @@
+
+    public InternshipGetByIdDto getByIdInternship(long internShipId) {
+       Internship internship =  internshipRepository.findById(internShipId)
+               .orElseThrow(() -> new EntityNotFoundException("InternShip not found"));
```
можно добавить лог
`log.error("Internship with ID {} not found", internshipId);`

RuslanZinovyev replied:
Я вижу палец вверх, но я не вижу изменений в коде, метод orElseThrow() не содержит логгера во время выбрасывания исключения.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,97 @@
+    private final List<Filter<TeamMember, InternshipFilterDto>> roleFilter;
+    private final List<Filter<Internship, InternshipFilterDto>> statusFilter;
+
+    public InternshipCreatedDto createInternship(InternshipCreatedDto internShipCreatedDto) {
```
имя переменной должно быть `internshipCreatedDto`

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,97 @@
+
+    public InternshipUpdatedDto updateInternship(InternshipUpdatedDto internShipUpdatedDto) {
+        Internship internship = internshipRepository.findById(internShipUpdatedDto.getId())
+                .orElseThrow(EntityNotFoundException::new);
```
можно добавить логгер
`log.error("Internship with ID {} not found", updateInternship.getId());`

RuslanZinovyev replied:
Логгер должен быть добавлен внутрь orElseThrow в момент когда ты выбрсываешь исключение, зачем тебе логировать ошибку если бизнес логика продолжилась?

Darkselect replied:
Исправил

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+
+    private List<Internship> filterByStatus(List<Internship> internships, InternshipFilterDto filters) {
```
Можно сделать сразу:
 
```
 private List<Internship> filterByStatus(List<Internship> internships, InternshipFilterDto filters) {
    return statusFilter.stream()
            .filter(filter -> filter.isApplicable(filters))
            .flatMap(filter -> filter.apply(internships.stream(), filters))
            .toList();
}
```

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,97 @@
+        return internshipMapper.toCreatedDto(internshipRepository.save(internshipMapper.createInternship(internShipCreatedDto)));
+    }
+
+    public InternshipUpdatedDto updateInternship(InternshipUpdatedDto internShipUpdatedDto) {
```
Можно сделать метод транзакционным.

Darkselect replied:
Добавил

RuslanZinovyev commented on src/test/java/faang/school/projectservice/controller/InternshipControllerTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Test
+    public void testExistingInternshipCreating() {
+        InternshipCreatedDto internShipCreatedDto = InternshipCreatedDto.builder().build();
```
`internship `- это одно слово, не надо переменную называть `internShip`  мы добавляем кэмэл кейс в качестве делиметра если только хотим разделить разные слова

Darkselect replied:
В смысле а то что я смайлики ставлю и нажимаю resolve ?

RuslanZinovyev commented on src/test/java/faang/school/projectservice/controller/InternshipControllerTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Test
+    public void testExistingInternshipUpdating() {
+        InternshipUpdatedDto internShipUpdatedDto = InternshipUpdatedDto.builder().build();
```
тоже самое, имя класса написано верно, а переменная нет
`internshipUpdatedDto`

Darkselect replied:
Оставил смайлик видно ?

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 9.0)
Fixed all

With all tests.

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 10)
Fixed all

With all tests.

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 11)
Fixed Controller

With all tests.

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 12)
Fixed Controller

With all tests.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,62 @@
+        return internshipService.filterInternship(filterInternship);
+    }
+
+    @GetMapping("/internships")
```
Плохая практика использоваться в одном месте `/internship` и в другом `/internships` именование URL для всех эндпоинтов должно быть предсказумо.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @GetMapping("/internship/{internshipId}")
+    public InternshipGetByIdDto getInternshipById(@Valid @PathVariable @Positive long internshipId) {
```
`Valid `тут не нужен, `Validated `на уровне класса будет ответственен за валидацию `PathVariable`

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,62 @@
+
+    @PutMapping("/internship")
+    public InternshipUpdatedDto updateInternship(@Valid @RequestBody InternshipUpdatedDto internship) {
+        log.info("Updating internship '{}' by UserId #{}.", internship.getId(), internship.getCreatedBy());
```
нам не нужен #, если ты хочешь красиво отформатировать логер то просто добавь : после UserId

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/InternshipService.java:
```diff
@@ -0,0 +1,106 @@
+
+@Service
+@Slf4j
+@Valid
```
Зачем нам эта аннотация тут?

RuslanZinovyev commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -1,16 +1,21 @@
+
 @Component
 @RequiredArgsConstructor
 @Slf4j
```
Я вижу аннотацию для логера, но я не вижу его использование в классе.

Darkselect replied:
Этот класс создавал не я

Darkselect replied:
а это ? 

if (projectRepository.existsByOwnerUserIdAndName(ownerId, name)) {
log.error("Project '{}' with ownerId #{} already exists.", name, ownerId);
throw new NotUniqueProjectException(String.format("Project '%s' with ownerId #%d already exists.",
name, ownerId));
}

        log.info("Project '{}' with ownerId #{} unique and can be created.", name, ownerId);

Darkselect replied:
Но положил сюда свой метод

public void validateMentorPresenceInProjectTeam(InternshipCreatedDto internShipCreatedDto) {
        Long projectId = internShipCreatedDto.getProjectId();
        Project project = projectRepository.getProjectById(projectId);
        List<Team> teams = project.getTeams();
        List<Long> teamMembersId = teams.stream()
                .flatMap(team -> team.getTeamMembers().stream())
                .map(TeamMember::getId)
                .toList();

        TeamMember mentorId = internShipCreatedDto.getMentorId();

        if (!isMentorPresent(teamMembersId, mentorId.getId())) {
            log.error("Mentor with id #{} is not present in project team", mentorId.getId());
            throw new IllegalArgumentException("Mentor is not present in project team");
        }
    }

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 13)
Fixed Controller

With all tests.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/InternshipController.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @GetMapping("/internship/filter")
+    public List<InternshipFilterDto> getAllInternshipsByFilter(@RequestParam InternshipFilterDto filterInternship) {
```
Почему тут `RequestParam`, ты передаешь DTO тут должна быть аннотация `RequestBody`

Darkselect replied:
Исправил

Darkselect committed:
BJS2-40653 Стажировка 5/5 (Version 14)
Fixed Controller

With all tests.
